File: plugin/hooks/get-read-posttool-output.py
#!/usr/bin/env python3
"""
get-read-posttool-output.py - Unified PostToolUse hook for Read/Glob/Grep/WebFetch/WebSearch

TRIGGER: PostToolUse (matcher: Read|Glob|Grep|WebFetch|WebSearch)

Consolidates read operation validation hooks into a single Python dispatcher.

Handlers can:
- Warn about patterns (return {"warning": "..."})
- Allow silently (return None)

Benefits:
1. Single Python startup vs N shell startups
2. Native JSON parsing (no jq subprocesses)
3. Shared state management
4. Better error handling
"""

import json
import sys
from pathlib import Path

# Add handlers directory to path
SCRIPT_DIR = Path(__file__).parent
sys.path.insert(0, str(SCRIPT_DIR))

from read_posttool_handlers import get_all_handlers


def read_hook_input() -> dict:
    """Read and parse JSON input from stdin."""
    if sys.stdin.isatty():
        return {}

    try:
        raw = sys.stdin.read()
        return json.loads(raw) if raw.strip() else {}
    except json.JSONDecodeError:
        return {}


def main():
    # Read hook input
    hook_data = read_hook_input()

    # Extract tool info
    tool_name = hook_data.get("tool_name", "")
    if tool_name not in ("Read", "Glob", "Grep", "WebFetch", "WebSearch"):
        print("{}")
        return

    tool_input = hook_data.get("tool_input", {})
    tool_result = hook_data.get("tool_result", {}) or hook_data.get("tool_response", {})

    # Build context for handlers
    context = {
        "tool_input": tool_input,
        "tool_result": tool_result,
        "session_id": hook_data.get("session_id", ""),
        "hook_data": hook_data,
    }

    # Run all handlers
    warnings = []
    for handler in get_all_handlers():
        try:
            result = handler.check(tool_name, context)
            if result and result.get("warning"):
                warnings.append(result["warning"])
        except Exception as e:
            sys.stderr.write(f"get-read-posttool-output: handler error: {e}\n")

    # Output warnings if any
    if warnings:
        for warning in warnings:
            sys.stderr.write(warning + "\n")

    # Always allow (PostToolUse cannot block, only warn)
    print("{}")


if __name__ == "__main__":
    main()

File: plugin/hooks/src/io/github/cowwoc/cat/hooks/GetReadPosttoolOutput.java
package io.github.cowwoc.cat.hooks;

import static io.github.cowwoc.requirements13.java.DefaultJavaValidators.requireThat;

import io.github.cowwoc.cat.hooks.read.post.DetectSequentialTools;
import tools.jackson.databind.JsonNode;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

/**
 * get-read-posttool-output - Unified PostToolUse hook for Read/Glob/Grep/WebFetch/WebSearch
 *
 * TRIGGER: PostToolUse (matcher: Read|Glob|Grep|WebFetch|WebSearch)
 *
 * Consolidates read operation validation hooks into a single Java dispatcher.
 *
 * Handlers can:
 * - Warn about patterns (return warning)
 * - Allow silently (return null)
 */
public final class GetReadPosttoolOutput
{
  private static final List<ReadHandler> HANDLERS = List.of(
      new DetectSequentialTools());

  private static final Set<String> SUPPORTED_TOOLS = Set.of(
      "Read", "Glob", "Grep", "WebFetch", "WebSearch");

  private GetReadPosttoolOutput()
  {
    // Utility class
  }

  /**
   * Entry point for the Read posttool output hook.
   *
   * @param _args command line arguments (unused)
   */
  @SuppressWarnings("UnusedVariable")
  public static void main(String[] _args)
  {
    HookInput input = HookInput.readFromStdin();

    String toolName = input.getToolName();
    if (!SUPPORTED_TOOLS.contains(toolName))
    {
      HookOutput.empty();
      return;
    }

    JsonNode toolInput = input.getToolInput();
    JsonNode toolResult = input.getToolResult();
    String sessionId = input.getSessionId();
    requireThat(sessionId, "sessionId").isNotBlank();
    List<String> warnings = new ArrayList<>();

    // Run all read posttool handlers
    for (ReadHandler handler : HANDLERS)
    {
      try
      {
        ReadHandler.Result result = handler.check(toolName, toolInput, toolResult, sessionId);
        // PostToolUse cannot block, only warn
        if (!result.reason().isEmpty())
          warnings.add(result.reason());
      }
      catch (Exception e)
      {
        System.err.println("get-read-posttool-output: handler error: " + e.getMessage());
      }
    }

    // Output warnings if any
    for (String warning : warnings)
    {
      System.err.println(warning);
    }

    // Always allow (PostToolUse cannot block, only warn)
    HookOutput.empty();
  }
}
